### PR TITLE
[ao] Fix to InputWeightEqualization detector to handle Conv groups

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81975
* #81974
* #81973
* #81673
* #81972
* __->__ #81971

Summary: The current implementation of the InputWeightEqualization
detector broke when it was tested on MobileNetV2, and the reason for
this is that it wasn't able to properly handle groups in Conv layers,
and there also had to be some minor reshaping of the weights to handle
this as well.

In addition, the output was correspondingly tuned so that instead of
giving on output for each channel on each layer, it gives a single
suggestion per module and just lets it know how many of the channels
could benefit from input-weight equalization, and suggests it if it's
more than half.

There was also the realization that the test class didn't do a good job
of testing different dimensions for the batch vs. height vs. width, so
this was updated to be more comprehensive as well.

Test Plan: python test/test_quantization TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 1]
```diff
@@ -789,10 +807,22 @@ def _generate_comparision_values(self, input_info: Dict, weight_info: Dict) -> D
             weight_ratio = self._calculate_range_ratio(weight_info[module_fqn], self.WEIGHT_STR, module_fqn)
             input_ratio = self._calculate_range_ratio(input_info[module_fqn], self.INPUT_STR, module_fqn)
 
+            # if mismatched size, because of grouping, we want to replicate weight enough times
+            weight_channels = len(weight_ratio)
+            input_channels = len(input_ratio)
+            if weight_channels != input_channels:
+                # we try to replicate
+                if input_channels % weight_channels == 0:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:815]: nit: can do `assert condition, msg` here

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 2]
```diff
@@ -789,10 +807,22 @@ def _generate_comparision_values(self, input_info: Dict, weight_info: Dict) -> D
...
+                    # get replication factor
+                    rep_factor: int = input_channels // weight_channels
+                    weight_ratio = weight_ratio.repeat(rep_factor)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:818]: which dimension is this repeating at? I think we might need some comments since this op is not very familiar to most of people

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 3]
```diff
@@ -910,15 +940,19 @@ def generate_detector_report(self, model: GraphModule) -> Tuple[str, Dict[str, A
...
-            # look at each individual channel and add a suggestion
-            for index, channel_suggested in enumerate(mod_info[self.RECOMMENDED_KEY]):
-                if channel_suggested:
-                    channel_str = channel_suggestion_str.format(index, use_str, input_weight_benefit_str)
-                    input_weight_string += channel_str
-                else:
-                    non_benefit_str = input_weight_non_benefit_str.format(input_weight_non_benefit_reasoning)
-                    channel_str = channel_suggestion_str.format(index, no_use_str, non_benefit_str)
-                    input_weight_string += channel_str
+            # gather info on how many channels would benefit from input weight and
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:943]: nit: would be better to have this in a separate PR since it's not related to groups handling

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 4]
```diff
@@ -789,10 +807,22 @@ def _generate_comparision_values(self, input_info: Dict, weight_info: Dict) -> D
             weight_ratio = self._calculate_range_ratio(weight_info[module_fqn], self.WEIGHT_STR, module_fqn)
             input_ratio = self._calculate_range_ratio(input_info[module_fqn], self.INPUT_STR, module_fqn)
 
+            # if mismatched size, because of grouping, we want to replicate weight enough times
+            weight_channels = len(weight_ratio)
+            input_channels = len(input_ratio)
+            if weight_channels != input_channels:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:813]: I'm wondering if we should be checking that the module is Conv2d here, otherwise it might be accepting cases that we do not want to handle

**Review by jerryzh168 — APPROVED** [review]
lg overall, please see some comments inline

File: torch/ao/quantization/fx/_model_report/detector.py (hunk 4 above)
Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/detector.py:813]: I think this mitigated by the fact that above, only linears are convs are listed as supported modules, so these are the only modules we could even be looking at in this case, so i think we're safe here.

File: torch/ao/quantization/fx/_model_report/detector.py (hunk 3 above)
Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/detector.py:943]: i think because there were minor changes made to the tests as well, i had to tweak this just a bit. I'll keep this in mind for future PRs
